### ISSUE TITLE
[Bug] : postcss-discard-duplicates crashes with TypeError: Cannot read property 'length' of undefined

### ISSUE THREAD

**Issue by Dema** [cssnano/cssnano#1031]
**Describe the bug**
I'm not sure what causes this error, I'm using Next.js 10.0.9, with bundled css-minimizer-webpack-plugin and postcss-nested, postcss-custom-media and I'm having this error

```
TypeError: Cannot read property 'length' of undefined
    at equals (/home/dema/Sync/work/t/OKTA/mashroom/main/node_modules/cssnano-preset-simple/dist/index.js:11834:42)
```
This happens in this line https://github.com/cssnano/cssnano/blob/master/packages/postcss-discard-duplicates/src/index.js#L58
```javascript
  if (a.nodes) {
    if (a.nodes.length !== b.nodes.length) {
      return false;
    }
```
Somehow `b.nodes` is undefined. Is it possible to add a condition like
```javascript
  if (a.nodes) {
    if (!b.nodes || a.nodes.length !== b.nodes.length) {
      return false;
    }
```
?

**Expected behavior**
I can't make a small reproducible case.

**Comment by alexander-akait** [comment]
> I can't make a small reproducible case.

We can't change code in this case

**Comment by Dema** [comment]
Can you give some advice on what is in `b.nodes`, so I could try to find out why it is undefined? I've tried to add console.log's ant there's some gigantic css file in `.source` field of both `a` and `b`.

The only differents in a and b is nodes:
![image](https://user-images.githubusercontent.com/207726/112726942-fd8cd700-8f30-11eb-8ddd-ef93812cb756.png)

Right now I can't reproduce it with smaller css, it just works. :disappointed:

**Comment by ludofischer** [comment]
@Dema Try looking at the PostCSS API https://postcss.org/api/#rule

**Comment by Dema** [comment]
For now, I'm just patching the code in-place in postinstall :)

I just was under impression that `equals` should be associative and if the left side can have `.nodes === undefined` so can the right side.

 Although here https://github.com/postcss/postcss/blob/a0a2d3954426140bd0db38dca5e12e44f5f3fde1/lib/container.d.ts#L41 is `nodes` is **not nullable**, so I'm a bit confused where the real error is — in postcss or in a plugin..

**Comment by sigveio** [comment]
Closing this since there has not been activity for a while. If there is still something unresolved, feel free to re-open. =) Ideally with link to a minimal reproducible case

**Comment by Dema** [comment]
Could you just add that condition? It's pretty obvious that you are comparing two objects of the same shape and for some reason you are testing `a.node !== undefined`, but do not test the `b.nodes !== undefined`. All equality checks should be commutative, if `a == b`, then `b == a` should not throw an exception.

**Comment by alexander-akait** [comment]
Please provide example of the problem, changing something without test is bad practice

**Comment by Dema** [comment]
I can't reproduce it using cssnano, but it fails in current next.js. Minimal code is something like

```
.z {
  color: red;
  @media (max-width: 500px);
}
```
this body-less media query has `nodes === undefined`

**Comment by alexander-akait** [comment]
it is invalid CSS...

**Comment by Dema** [comment]
Yes, I agree, but it shouldn't cause plugin to crash with `Cannot read property 'length' of undefined`. And docs says that `nodes` can be undefined...  

Worst thing is this seems to be a **valid** css. Even stylelint doesn't complain about it. I tried in their playground.

Yep, if I delete this invalid media query it doesn't crash anymore.

**Comment by Dema** [comment]
Thank you very much!!!

**Comment by sigveio** [comment]
Thanks for adding additional info, @Dema 👍 

> Could you just add that condition? It's pretty obvious (...)

We ask for something to reproduce with because it not only makes it a lot easier to assess/confirm whether it is in fact a bug or not, but also as @alexander-akait pointed out - to be able to construct tests and know that we are fixing the problem. And it generally saves a lot of "back and forth" and time for everyone involved :-)

**Comment by SmallhillCZ** [comment]
> Using invalid CSS syntax can break minimizer

the mentioned CSS might be invalid, but the same error pops up when using `@layer b;` which is valid CSS according to https://developer.mozilla.org/en-US/docs/Web/CSS/@layer

see also https://github.com/vercel/next.js/issues/68476

**Comment by SmallhillCZ** [comment]
Ok nevermind, it is probably fixed in https://github.com/cssnano/cssnano/issues/1655
